sled-agent/src/artifact_store.rs

```diff
@@ -0,0 +1,481 @@
+        }
+
+        let moved_final_path = final_path.clone();
+        tokio::task::spawn_blocking(move || {
+            file.persist(&moved_final_path).map_err(|err| Error::File {
+                verb: "rename temporary file to",
+                path: moved_final_path,
+                err: err.error,
+            })
+        })
+        .await??;
+        info!(
+            &self.log,
+            "Wrote artifact";
+            "sha256" => &sha256.to_string(),
+            "path" => final_path.as_str()
+        );
+        Ok(())
+    }
+
+    /// PUT operation (served by Sled Agent API) which takes a [`StreamingBody`]
+    pub(crate) async fn put_body(
+        &self,
+        sha256: ArtifactHash,
+        body: StreamingBody,
+    ) -> Result<(), Error> {
+        self.put(sha256, body.into_stream().map_err(Error::Body)).await
+    }
+
+    /// POST operation (served by Sled Agent API)
+    pub(crate) async fn copy_from_depot(
+        &self,
+        sha256: ArtifactHash,
+        address: SocketAddrV6,
+    ) -> Result<(), Error> {
+        static CLIENT: LazyLock<reqwest::Client> = LazyLock::new(|| {
+            reqwest::ClientBuilder::new()
+                .connect_timeout(Duration::from_secs(15))
+                .read_timeout(Duration::from_secs(15))
+                .build()
+                .unwrap()
+        });
+
+        let client = repo_depot_client::Client::new_with_client(
+            &format!("http://{address}"),
+            CLIENT.clone(),
+            self.log.new(
+                slog::o!("component" => "Repo Depot client (ArtifactStore)"),
+            ),
+        );
+        let stream = client
+            .artifact_get_by_sha256(&sha256.to_string())
+            .await
+            .map_err(|err| Error::DepotCopy { sha256, address, err })?
+            .into_inner()
+            .into_inner()
+            .map_err(|err| Error::DepotCopy {
+                sha256,
+                address,
+                err: repo_depot_client::ClientError::ResponseBodyError(err),
+            });
+        self.put(sha256, stream).await
+    }
+
+    /// DELETE operation (served by Sled Agent API)
+    pub(crate) async fn delete(
+        &self,
+        sha256: ArtifactHash,
+    ) -> Result<(), Error> {
+        let sha256 = sha256.to_string();
+        let mut any_datasets = false;
+        for dataset in self.datasets().await? {
+            any_datasets = true;
+            let path = dataset.join(&sha256);
+            match tokio::fs::remove_file(&path).await {
+                Ok(()) => {
+                    info!(
+                        &self.log,
+                        "Removed artifact";
+                        "sha256" => &sha256,
+                        "path" => path.as_str()
+                    );
+                }
+                Err(err) if err.kind() == ErrorKind::NotFound => {}
+                Err(err) => {
+                    return Err(Error::File { verb: "remove", path, err });
```

Comment:
Same here -- we should probably keep trying across other datasets.

My fear (maybe rational, maybe not -- feedback welcome) is that we try to access a dataset from a disk that has been yanked from the system. Let's suppose we get non-stop I/O error accessing paths within this dataset mountpoint --- but an update dataset on a **different** disk still has this artifact. We should still try that second dataset, even if the first fails.

Reply:
I think that makes sense. The flow I am considering is:

- get: on non-`NotFound`, log the error. Return the artifact if it's found on another dataset, or return the most recently-encountered error (instead of a NotFound).
- delete: on non-`NotFound`, log the error. If any errors occurred return the most recently-encountered error (because deleting all copies of that particular artifact was not successful).

Reply:
makes sense to me. For `GET`, it's reasonable to return early if we find any valid dataset, but for `DELETE`, we should probably look through all datasets to see if it exists on any of them, and try to delete all of them.

(I know that the plan is to only have one "artifact store" per sled, so this is a bit future-proofy, but I think it would minimize weirdness)